### ISSUE TITLE
How to extract the gradient of the generator in one-stage

### ISSUE THREAD

**Issue by naoki7090624** [zju-vipa/OSGAN#1]
Thanks for sharing your code and it's really a great work.

I would like to apply this technique to other models using GANs to reduce the training time.
For implementation, I have two questions about your code.

1. Why do you set ```reduction='none'``` and get the average from ```loss_fake``` in one stage ?
[train_dcgan_asymmetric_one_stage_simple.py#L60](https://github.com/zju-vipa/OSGAN/blob/main/train_dcgan_asymmetric_one_stage_simple.py#L60)
2. How do you extract the gradient of the generator from the gradient of the Discriminator?
In this code [train_dcgan_asymmetric_one_stage_simple.py#L72](https://github.com/zju-vipa/OSGAN/blob/main/train_dcgan_asymmetric_one_stage_simple.py#L72), I think that the generator backpropagates the same gradient as the discriminator.

Thank you.

**Comment by sccbhxc** [comment]
Thanks for your watching on our work.
1. Since we need image-wise loss in the following function: `get_gradient_ratios()` instead of batch average loss to obtain image-wise gradient ratio (`gamma` is a vector, whose size is equal to batch size of fake images), we set `reduction='none` in `loss_fake` and `loss_g`. 
2. We scale the gradients derived from generator as the gradients of discriminator. The function is achieved in `GradientScaler` : https://github.com/zju-vipa/OSGAN/blob/fdfea7e7d9c7627d19fca26b992aaac502a02313/utils/modules.py#L4, which corresponds to Eq.S8 in supplemental material: https://openaccess.thecvf.com/content/CVPR2021/supplemental/Shen_Training_Generative_Adversarial_CVPR_2021_supplemental.pdf.

**Comment by naoki7090624** [comment]
Thanks for your reply.


Does the ```GradientScaler``` function only scale the gradient of the generator and not the Discriminator?

Sorry for the additional question. Can OSGAN be used even if there are losses other than adversarial loss? For example, generator loss = adversarial loss + perceptual loss + reconstruction loss.

**Comment by ElegantLee** [comment]
> Thanks for your reply.
> 
> Does the `GradientScaler` function only scale the gradient of the generator and not the Discriminator?
> 
> Sorry for the additional question. Can OSGAN be used even if there are losses other than adversarial loss? For example, generator loss = adversarial loss + perceptual loss + reconstruction loss.

I have the same question about how to do back propagation when we have more than adversarial loss(Generator have more than adv loss).  The way I solved this problem is: loss_pack = loss_pack + other g losses. I'm not sure if this method is right. I hope the author can answer it. :joy:
